stdlib/public/TensorFlow/CompilerRuntime.swift

```diff
@@ -87,10 +87,195 @@ public enum _ExecutionMode : Equatable {
   }
 }
 
+// TraceContext contains the state as needed to build a trace graph function
+// (TF_Function). As eager ops are executed in tracing mode, their corresponding
+// nodes are added to the trace graph (via `addEagerOpToGraph()`). When the
+// trace is fully formed (via `finalizeAndExecute()`), the trace graph function
+// is run by the eager runtime.
+private class TraceContext {
+  let status: CTFStatus = TF_NewStatus()
+
+  // The trace graph, which will be converted to a trace graph function
+  // (TF_Function) upon finalizing.
+  let graph = TF_NewGraph()
+
+  // The list of inputs to the trace graph function.
+  //
+  // These symbolic tensors corresond to PlaceHolder nodes in the trace graph,
+  // and will be filled in when we execute the trace graph function.
+  //
+  // TODO: If some tensors in `x` are not used within `foo()`, they can be
+  // pruned away in the inputs to the trace graph function.
+  var symbolicInputs: [TF_Output] = []
+
+  // The trace context object used in TF C API calls that convert eager ops to
+  // (trace) graph nodes.
+  let cTraceContext: CTFETraceContext
+
+  // `inputValueCount` is the length of the (flattened) list of input tensors to
+  // the trace function.
+  init(inputValueCount: Int) {
+    debugLog("Instiantiating TraceContext with \(inputValueCount) input tensors.")
+    for i in 0..<inputValueCount {
+      let desc = TF_NewOperation(graph, "Placeholder", "input_\(i)")
+      // TODO: add other dtype support.
+      TF_SetAttrType(desc, "dtype", TF_FLOAT)
+      let result = TF_FinishOperation(desc, status)
+      checkOk(status)
+      symbolicInputs.append(TF_Output(oper: result, index: 0))
+    }
+    traceGraphObjectCounter = inputValueCount
+
+    cTraceContext = TFE_NewTraceContext(graph)
+  }
+
+  deinit {
+    TFE_DeleteTraceContext(cTraceContext)
```

Comment:
nit, does the order of destruction matter? TraceContext seems to hold onto graph?

Reply:
Trace graph is created first, and then fed into a trace context object (which does not own the graph). So the destruction should be in the reverse ordering.